Implement Cross-Signing

TobiasFella opened the pull request:
The parts involved are roughly

- When receiving device updates:
	- Master Key: Check for validity and save
	- Self Signing Key: Check for validity and signature, then save
	- User signing Key: Check for validity and signature, then save
	- Device Keys: Check if they're signed by the user's SSK, if so, mark it as self-verified
	- If we trust our own master key:
	    - Check if our USK signs any other user's master keys, if so, mark them as verified
	- If a user's master key changes:
	    - Mark devices for that user as unverified and delete SSK for the user

- When checking if a megolm session is verified:
	- Check if the device is verified manually, if so, return true - Check if a device is marked as self-verified AND the master key is verified, if so, return true - Else return false

- When verifying a device - Mark the other party's master key as verified in addition to the device - We can't sign the master key yet, since we don't have the USK

TobiasFella committed:
Implement Cross-Signing

The parts involved are roughly

- When receiving device updates:
	- Master Key: Check for validity and save
	- Self Signing Key: Check for validity and signature, then save
        - User signing Key: Check for validity and signature, then save
        - Device Keys: Check if they're signed by the user's SSK, if so, mark it as self-verified

        - If we trust our own master key:
		Check if our USK signs any other user's master keys, if so, mark them as verified
        - If a user's master key changes:
		Mark devices for that user as unverified and delete SSK for the user

- When checking if a megolm session is verified:
	- Check if the device is verified manually, if so, return true
        - Check if a device is marked as self-verified AND the master key is verified, if so, return true
        - Else return false

- When verifying a device
        - Mark the other party's master key as verified in addition to the device
        - We can't sign the master key yet, since we don't have the USK

TobiasFella commented on lib/keyverificationsession.cpp:
```diff
@@ -379,12 +379,21 @@ void KeyVerificationSession::handleMac(const KeyVerificationMacEvent& event)
         return;
     }
 
+    auto masterKey = m_connection->masterKeyForUser(m_remoteUserId);
```
(Note to myself:)

We need to distinguish between device ids and master keys during verification

TobiasFella replied:
done

KitsuneRal commented on lib/connection.cpp:
```diff
@@ -1884,7 +2046,11 @@ void Connection::Private::loadOutdatedUserDevices()
         currentQueryKeysJob = nullptr;
         if (queryKeysJob->error() == BaseJob::Success) {
-            handleQueryKeys(queryKeysJob);
+            QFile f("/home/tobias/keys"_ls);
```
Don't forget to fix that :)))

KitsuneRal commented on autotests/testcrosssigning.cpp:
```diff
@@ -0,0 +1,45 @@
+
+
+#include <QTest>
+#define private public
+#include "connection_p.h"
+#undef private
```
Do you really need this hack? I don't see any private members in `connection_p.h`.

TobiasFella replied:
It's not about private members in connection_p, it's about `d` being private in Connection

TobiasFella commented:
Why does github not run CI now?!

TobiasFella commented:
ping :)

KitsuneRal commented:
Can you rebase this to `dev`? Hopefully CI acts saner now.

TobiasFella commented:
Didn't test the rebase yet; should also be cleaned up a bit

KitsuneRal commented:
The rebase superficially looks fine for E2EE but tries to call into E2EE pieces in non-E2EE configuration, failing it.

KitsuneRal commented:
Also, I seem to have just added a conflict again with the `QByteArray/QString` PR merged :( sorry.

KitsuneRal commented:
I think know why linking likely fails on Windows - because `struct CrossSigningKeys` is not annotated with `QUOTIENT_API`. Gotta change the template for top-level structures generation (`data.h.mustache`, that is).

KitsuneRal commented:
I think I have another guess of what is happening here - it might be that the way precompiled headers building is implemented in MSVC is not compatible with the private/public hack that you're using in the test. I would suggest looking into `friend`ing `TestCrossSigning` in `Connection` and `ConnectionEncryptionData` for now, as a much less intrusive hack. Actually, with regards to `ConnectionEncryptionData`, I'd just pull `handleQueryKeys()` out to `public`; that  class is already "private" from the client code, anyway.

And then you won't need the static building condition, either.

vkrause commented:
> I think I have another guess of what is happening here - it might be that the way precompiled headers building is implemented in MSVC is not compatible with the private/public hack that you're using in the test. 

It's not even limited to precompiled headers, the access specifier is part of the symbol name on Windows, so that define hack cannot ever work there.

KitsuneRal commented:
> > I think I have another guess of what is happening here - it might be that the way precompiled headers building is implemented in MSVC is not compatible with the private/public hack that you're using in the test.
> 
> It's not even limited to precompiled headers, the access specifier is part of the symbol name on Windows, so that define hack cannot ever work there.

Ah, that certainly explains why it's happening on Windows.

KitsuneRal commented on Quotient/csapi/definitions/cross_signing_key.h:
```diff
@@ -8,7 +8,7 @@
 namespace Quotient {
 /// Cross signing key
-struct CrossSigningKey {
+struct QUOTIENT_API CrossSigningKey {
```
You have to do it in mustache. Honestly, it would be the right thing to add this macro to all generated code, as some structures may be useful for clients. I'll make a separate PR for that and you will rebase yours once that other one  is merged, ok?

TobiasFella replied:
yeah, sure

redstrate commented on Quotient/connection.cpp:
```diff
@@ -1882,6 +1911,20 @@ bool Connection::isKnownE2eeCapableDevice(const QString& userId, const QString&
+void Connection::reloadDevices()
+{
+    if (d->encryptionData) {
+        reloadDevices();
```
I seem to get an infinite loop here sometimes, and I can't really tell what it's supposed to be doing. Seems to be a rebasing error?

TobiasFella replied:
yeah this should probably be d->encryptionData->reloadDevices()

redstrate replied:
Can confirm, changing it to this prevents an infinite loop. Easiest way to tell is when you complete the verification on the other end before in NeoChat, because NeoChat will freeze :laughing:

KitsuneRal replied:
```suggestion
        d->encryptionData->reloadDevices();
```

KitsuneRal commented on Quotient/connectionencryptiondata_p.cpp:
```diff
@@ -371,26 +376,183 @@ void ConnectionEncryptionData::handleEncryptedToDeviceEvent(
             }
             if (!verifyIdentitySignature(device, device.deviceId,
-                                         device.userId)) {
+                                        device.userId)) {
```
```suggestion
                                         device.userId)) {
```

KitsuneRal commented on Quotient/connectionencryptiondata_p.cpp:
```diff
@@ -371,26 +376,183 @@ void ConnectionEncryptionData::handleEncryptedToDeviceEvent(
             }
             if (!std::all_of(device.algorithms.cbegin(),
-                             device.algorithms.cend(), isSupportedAlgorithm)) {
+                            device.algorithms.cend(),
+                            isSupportedAlgorithm)) {
```
It seems that your editor disregards `.clang-format`.
```suggestion
                             device.algorithms.cend(), isSupportedAlgorithm)) {
```

KitsuneRal commented on Quotient/connectionencryptiondata_p.cpp:
```diff
@@ -422,7 +592,7 @@ void ConnectionEncryptionData::handleQueryKeys(const QueryKeysJob* job)
                       if (!isKnownCurveKey(
                               pendingEvent->fullJson()[SenderKey].toString(),
-                              pendingEvent->contentPart<QString>(SenderKeyKey)))
+                              pendingEvent->contentPart<QString>(SenderKey)))
```
That can't be correct, can it? `SenderKeyKey == "sender_key"_ls` and `SenderKey == "sender"_ls`.
```suggestion
                              pendingEvent->contentPart<QString>(SenderKeyKey)))
```

TobiasFella replied:
Fixed in a nicer way

KitsuneRal commented on Quotient/keyverificationsession.cpp:
```diff
@@ -55,6 +55,10 @@ KeyVerificationSession::KeyVerificationSession(
     , m_remoteSupportedMethods(event.methods())
 {
+    if (connection->hasConflictingDeviceIdsAndCrossSigningKeys(m_remoteUserId)) {
+        qWarning() << "Remote user has conflicting device ids and cross signing keys and device ids; refusing to verify.";
```
There seems to be one "and" too much in this sentence.

KitsuneRal commented on Quotient/keyverificationsession.cpp:
```diff
@@ -73,6 +77,10 @@ KeyVerificationSession::KeyVerificationSession(QString userId, QString deviceId,
     , m_encrypted(false)
 {
+    if (connection->hasConflictingDeviceIdsAndCrossSigningKeys(m_remoteUserId)) {
+        qWarning() << "Remote user has conflicting device ids and cross signing keys and device ids; refusing to verify.";
```
Ditto.

KitsuneRal commented on autotests/cross_signing_data.json:
```diff
@@ -0,0 +1,926 @@
+{
+  "device_keys": {
```
Do you need that many keys in the test data?

Also - I assume that `@aloy:kde.org`, Carl and you wilfully exposed your data, including device display names? This is not something normally shown to other users.

TobiasFella replied:
Right, i removed most devices. I randomized the display names a bit. For the rest, i'm personally fine with it, since it's not secret data. I'll ping @CarlSchwan  whether he's fine with it. `@aloy:kde.org` is my test account :)

TobiasFella replied:
carl mentioned he's fine with it

KitsuneRal commented on autotests/CMakeLists.txt:
```diff
@@ -26,4 +26,9 @@ if(${PROJECT_NAME}_ENABLE_E2EE)
     quotient_add_test(NAME testolmutility)
     quotient_add_test(NAME testfilecrypto)
     quotient_add_test(NAME testkeyverification)
+    if(${BUILD_SHARED_LIBS})
+        message("Cross-Signing test requires building libQuotient statically")
+    else()
+        quotient_add_test(NAME testcrosssigning)
+    endif()
```
That conditional should not be needed any more.
```suggestion
    quotient_add_test(NAME testcrosssigning)
```

KitsuneRal reviewed (changes requested):
I'm afraid it needs another even closer look...

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -17,6 +17,7 @@
 
 // NB: since Qt 6, moc_connection.cpp needs Room and User fully defined
 #include "moc_connection.cpp" // NOLINT(bugprone-suspicious-include)
+#include "util.h"
```
Not needed, came from a clumsy conflict resolution.
```suggestion
```

TobiasFella committed:
Various improvements and fixes

TobiasFella committed:
Assorted fixes and hacks

- User verification now mostly works for incoming requests
- Outgoing user verification is still missing
- We're trying to sign the other user's master key, but failing so far

TobiasFella committed:
More assorted fixed

- We're now signing other user's master keys correctly
- CS keys and backup are now loaded automatically after verifying our own device

TobiasFella committed:
Self-signing also works :)

TobiasFella committed:
Implement outgoing user verification

TobiasFella committed:
Things

KitsuneRal commented on Quotient/connectionencryptiondata_p.cpp:
```diff
@@ -382,12 +387,168 @@ void ConnectionEncryptionData::handleEncryptedToDeviceEvent(
+
+void ConnectionEncryptionData::handleQueryKeys(const QHash<QString, QHash<QString, QueryKeysJob::DeviceInformation>>& deviceKeys,
+                     const QHash<QString, CrossSigningKey>& masterKeys, const QHash<QString, CrossSigningKey>& selfSigningKeys,
+                     const QHash<QString, CrossSigningKey>& userSigningKeys)
```
This function is exceptionally long, and together with the generic name, it makes rather hard to grasp what it actually does. Any chance of splitting it into several self-contained parts with somewhat more specific names?

(Just to give a feeling - as much as Sonar's cognitive complexity metric is inaccurate and so, it goes about 3 times higher than the recommended maximum.)

TobiasFella replied:
sure

TobiasFella replied:
done

KitsuneRal commented on Quotient/e2ee/cryptoutils.cpp:
```diff
@@ -320,3 +320,20 @@ std::vector<byte_t> Quotient::base58Decode(const QByteArray& encoded)
     return result;
 }
+
+QUOTIENT_API QByteArray Quotient::sign(const QByteArray& key, const QByteArray& data)
```
```suggestion
QByteArray Quotient::sign(const QByteArray& key, const QByteArray& data)
```
It's enough to say `QUOTIENT_API` in the header file.

KitsuneRal commented on Quotient/e2ee/cryptoutils.cpp:
```diff
@@ -320,3 +320,20 @@ std::vector<byte_t> Quotient::base58Decode(const QByteArray& encoded)
+    const auto signatureLength = olm_pk_signature_length();
+    auto signatureBuffer = byteArrayForOlm(signatureLength);
+
+    if (olm_pk_sign(context.get(), (const uint8_t *) data.data(), unsignedSize(data), (uint8_t*) signatureBuffer.data(), signatureLength)
```
```suggestion
    if (olm_pk_sign(context.get(), asCBytes(data).data(), unsignedSize(data), asWritableCBytes(signatureBuffer).data(), signatureLength)
```

KitsuneRal commented on Quotient/keyverificationsession.cpp:
```diff
@@ -64,6 +72,35 @@ KeyVerificationSession::KeyVerificationSession(
     // Otherwise don't even bother starting up
```
I have an uneasy feeling that we are leaking objects in these constructors when we return without setting timeouts or sending an outgoing requests (not just in this PR but since "Don't bother starting up" comment emerged). Not completely leaking, they are still formally under `Connection`; but those objects end up stuck in `d->encryptionData->verificationSessions` until the end of `Connection` lifetime.

TobiasFella replied:
right - especially with the "send key verification requests to all of our own devices" thing that's also in this MR, we'll need to come up with  a way of cleaning them up.

TobiasFella replied:
I've added a `deleteLater` in those cases, together with a connection to `QObject::destroyed` that removes the connection from the list. (which already was there for in the existing logic, I just added it to the new cases)

KitsuneRal commented on Quotient/room.cpp:
```diff
@@ -2844,6 +2848,40 @@ Room::Changes Room::Private::addNewMessageEvents(RoomEvents&& events)
+        }
+        if (const auto* evt = it->viewAs<RoomMessageEvent>()) {
+            if (evt->rawMsgtype() == "m.key.verification.request"_ls) {
+                auto session = new KeyVerificationSession(evt->senderId(), connection, q->usesEncryption(), {}, evt->contentPart<QStringList>("methods"_ls), evt->originTimestamp(), evt->contentPart<QString>("from_device"_ls), q, evt->id());
```
I think we can still pass the entire event into the constructor and handle the differences between room and to-device events inside the session constructor.

KitsuneRal commented on Quotient/keyverificationsession.cpp:
```diff
@@ -64,6 +72,35 @@ KeyVerificationSession::KeyVerificationSession(
+
+KeyVerificationSession::KeyVerificationSession(
+    QString remoteUserId,
+    Connection* connection, bool encrypted, const QString& transactionId, const QStringList& methods, const QDateTime& timestamp, const QString &deviceId, Room *room, const QString &requestEventId
```
If you pass a room pointer, the connection pointer is trivially resolved from it.

KitsuneRal commented on Quotient/keyverificationsession.cpp:
```diff
@@ -421,6 +509,70 @@ void KeyVerificationSession::handleMac(const KeyVerificationMacEvent& event)
+                // TODO: Delete the handler once it's finished
+                auto handler = new SSSSHandler(m_connection);
+                handler->setConnection(m_connection);
+                handler->unlockSSSSFromCrossSigning();
```
Does it make sense to `deleteLater()` upon successful unlocking of the session? I realise it's not trivial atm because of several network requests flying - I'm working on something that might help that :)

TobiasFella replied:
I'm considering to just *not* unlock SSSS automatically. that would clear up this problem a bit and would give the client more explicit control (and information of) the unlocking operation

KitsuneRal commented on Quotient/room.cpp:
```diff
@@ -2844,6 +2848,40 @@ Room::Changes Room::Private::addNewMessageEvents(RoomEvents&& events)
+                keyVerificationSessions[evt->id()] = session;
+            }
+        }
+        if (it->event()->matrixType().startsWith("m.key.verification."_ls)) {
+            auto event = loadEvent<KeyVerificationEvent>(it->event()->fullJson());
```
This really should not be so complicated.
```suggestion
        if (auto event = it->viewAs<KeyVerificationEvent>()) {
```
But make sure to derive `KeyVerificationEvent` from `RoomEvent` so that normal room event deserialisation could use these classes.

KitsuneRal reviewed (changes requested):
Next review round. I understand the challenge with `KeyVerificationEvent` being and not being a room event - at the very least, it doesn't hurt at all to derive it from `RoomEvent` (that still allows to deserialise it as a to-device event too). I'll look into abstracting other things away so that we don't have to massage JSON payloads etc.

TobiasFella committed:
Improve

TobiasFella committed:
Update Quotient/room.cpp

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
:)

TobiasFella committed:
Remember verified devices

TobiasFella committed:
Cleanup

TobiasFella committed:
Cleanup key verification sessions

TobiasFella committed:
Cleanup and error handling

TobiasFella committed:
Fixes

KitsuneRal commented on Quotient/connectionencryptiondata_p.h:
```diff
@@ -9,6 +9,8 @@
 #include "events/encryptedevent.h"
 
+#include "events/encryptedevent.h"
+
```
```suggestion
```
It's already there.

KitsuneRal commented on Quotient/e2ee/sssshandler.cpp:
```diff
@@ -217,6 +217,7 @@ void SSSSHandler::loadMegolmBackup(const QByteArray& megolmDecryptionKey)
                         static_cast<uint32_t>(backupData.firstMessageIndex));
                 }
             }
+            Q_EMIT finished();
```
```suggestion
            emit finished();
```

KitsuneRal commented on Quotient/events/keyverificationevent.h:
```diff
@@ -3,17 +3,17 @@
 #pragma once
 
-#include "event.h"
+#include <Quotient/events/roomevent.h>
```
```suggestion
#include "roomevent.h"
```

KitsuneRal commented on Quotient/room.cpp:
```diff
@@ -2845,6 +2849,45 @@ Room::Changes Room::Private::addNewMessageEvents(RoomEvents&& events)
+            }
+            if (keyVerificationSessions.contains(baseEvent)) {
+                keyVerificationSessions[baseEvent]->handleEvent(*event);
+            } else qWarning() << "unknown session";
```
```suggestion
            } else
                qCWarning(E2EE) << "Unknown verification session, id" << baseEvent;
```

KitsuneRal commented on Quotient/room.cpp:
```diff
@@ -2845,6 +2849,45 @@ Room::Changes Room::Private::addNewMessageEvents(RoomEvents&& events)
+                continue;
+            } else {
+                continue;
+            }
```
```suggestion
            if (const auto* evt = it->viewAs<RoomMessageEvent>()) {
                if (evt->rawMsgtype() == "m.key.verification.request"_ls && pendingKeyVerificationSession && evt->senderId() == q->localMember().id()) {
                    keyVerificationSessions[evt->id()] = pendingKeyVerificationSession;
                    connect(pendingKeyVerificationSession.get(), &QObject::destroyed, q, [this, evt] {
                        keyVerificationSessions.remove(evt->id());
                    });
                    pendingKeyVerificationSession->setRequestEventId(evt->id());
                    pendingKeyVerificationSession.clear();
                }
            }
            continue;
```

KitsuneRal commented on Quotient/keyverificationsession.cpp:
```diff
@@ -322,25 +392,33 @@ void KeyVerificationSession::sendReady()
 
     if (methods.size() == 1) {
-        sendStartSas();
+        //sendStartSas();
```
?

TobiasFella replied:
I think i was trying to debug some stability problem there... but i don't remember, what the conclusion was. I'll revert that. (it works fine with that line in)

KitsuneRal reviewed (commented):
Okay, I'm almost good to merge this as-is, except that commented out `sendStartSas()` call. Otherwise, agree it's about as good to go, and the rest can be ironed out by separate PRs.

TobiasFella committed:
un-comment out line

TobiasFella committed:
Update Quotient/room.cpp

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Update Quotient/room.cpp

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Apply suggestions from code review

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

TobiasFella committed:
Update Quotient/connectionencryptiondata_p.h

Co-authored-by: Alexey Rusakov <Kitsune-Ral@users.sf.net>

KitsuneRal reviewed (approved):
Let's go!
